src/openms/source/CHEMISTRY/TheoreticalSpectrumGeneratorNA.cpp

```diff
@@ -0,0 +1,554 @@
+      {
+        Size ribo_index = oligo.size() - i - 1;
+        fragments_right[i] = (fragments_right[i - 1] + ribo_masses[ribo_index] +
+                              backbone_mass);
+      }
+      if (add_w_ions_)
+      {
+        addFragmentPeaks_(spectrum, fragments_right, "w", w_ion_offset,
+                          w_intensity_);
+      }
+      if (add_x_ions_)
+      {
+        addFragmentPeaks_(spectrum, fragments_right, "x", x_ion_offset,
+                          x_intensity_);
+      }
+      if (add_y_ions_)
+      {
+        addFragmentPeaks_(spectrum, fragments_right, "y", y_ion_offset,
+                          y_intensity_);
+      }
+      if (add_z_ions_)
+      {
+        addFragmentPeaks_(spectrum, fragments_right, "z", z_ion_offset,
+                          z_intensity_);
+      }
+    }
+
+    if (add_precursor_peaks_) // re-use what we've already calculated
+    {
+      Peak1D peak(0.0, precursor_intensity_);
+      bool have_left = !fragments_left.empty();
+      bool have_right = !fragments_right.empty();
+      if (have_left && have_right)
+      {
+        peak.setMZ(fragments_left[0] + fragments_right.back() + backbone_mass);
+      }
+      else if (have_left)
+      {
+        peak.setMZ(fragments_left.back() + ribo_masses.back() + backbone_mass +
+                   three_prime_mass);
+      }
+      else if (have_right)
+      {
+        peak.setMZ(fragments_right.back() + ribo_masses[0] + backbone_mass +
+                   five_prime_mass);
+      }
+      else // really, no fragment ions?
+      {
+        peak.setMZ(oligo.getMonoWeight(NASequence::Full, 0));
+      }
+      spectrum.push_back(peak);
+      if (add_metainfo_)
+      {
+        spectrum.getStringDataArrays()[0].push_back("M");
+      }
+    }
+
+    return spectrum;
+  }
+
+
+  void TheoreticalSpectrumGeneratorNA::addChargedSpectrum_(
+    MSSpectrum& spectrum, const MSSpectrum& uncharged_spectrum, Int charge,
+    bool add_precursor) const
+  {
+    // based on very limited testing, it looks like this hydrogen mass gives
+    // better results than using "Constants::PROTON_MASS_U" (why?):
+    static const double H_mass = EmpiricalFormula("H").getMonoWeight();
+    if (uncharged_spectrum.empty()) return;
+    Size size = uncharged_spectrum.size();
+    if (add_precursor_peaks_ && !add_precursor)
+    {
+      --size; // uncharged spectrum contains precursor peak - exclude it
+    }
+    for (Size i = 0; i < size; ++i)
+    {
+      spectrum.push_back(uncharged_spectrum[i]);
+      double mass = spectrum.back().getMZ() + charge * H_mass;
+      spectrum.back().setMZ(abs(mass / charge));
+    }
+    if (add_metainfo_)
+    {
+      auto& ions = spectrum.getStringDataArrays()[0];
+      auto source_it = uncharged_spectrum.getStringDataArrays()[0].begin();
+      ions.insert(ions.end(), source_it, source_it + size);
+      auto& charges = spectrum.getIntegerDataArrays()[0];
+      charges.resize(charges.size() + size, charge);
```

Comment:
how do you know that `[0]` is the charge array?
Why resize based on old value (the number of elements should be exactly 'size')?

Reply:
Using the first array (`[0]`) is a convention from the original `TheoreticalSpectrumGenerator`. I don't really like it, either - see https://github.com/OpenMS/OpenMS/pull/4020/files/dcb360e2cf28073f5906e64fc7cd833aacd26e2b#diff-429a6a69257db40684324599fff0e7bcR390. I think it would have made more sense to use the last array. I didn't change the convention because I don't know what parts of OpenMS depend on it. At least it's documented: https://github.com/OpenMS/OpenMS/pull/4020/files/dcb360e2cf28073f5906e64fc7cd833aacd26e2b#diff-5408e877edfef997d52836ef8df272b7L58

This function *adds* peaks to a spectrum. So the `resize` extends the size of the array by the number of new peaks.